Apply new experimental metrics builder to hostmetrics/cpu scraper

dmitryax opened the pull request:
Introduce new experimental metrics builder interface for metrics scrapers and apply it to hostmetrics/cpu scraper.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/10904

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go:
```diff
@@ -14,84 +14,134 @@
-type metricStruct struct {
-	SystemCPUTime MetricIntf
+// MetricBuilderOption applies changes to default metrics builder.
+type MetricBuilderOption func(*MetricsBuilder)
```
you want the "func(*MetricsBuilder)" to work on something internal (private) so people cannot implement an option.

dmitryax replied:
made metricBuilderOption private

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go:
```diff
@@ -14,84 +14,134 @@
-	New() pdata.Metric
-	Init(metric pdata.Metric)
+// MetricConfig provides common settings for a particular metric.
+type MetricConfig struct {
```
We use "Settings" for this when ask users to embed into their config.

dmitryax replied:
Changed to Settings

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go:
```diff
@@ -14,84 +14,134 @@
-// Name returns the metric name.
-func (m *metricImpl) Name() string {
-	return m.name
+func CreateDefaultMetricsConfig() MetricsConfig {
```
Just `DefaultMetricsConfig`. or `NewDefaultMetricsConfig`. See equivalent https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L52

dmitryax replied:
done

bogdandrutu commented on cmd/mdatagen/metrics_v2.tmpl:
```diff
@@ -0,0 +1,203 @@
+
+// Code generated by mdatagen. DO NOT EDIT.
+
+// {{ .ExpFileNote }}
```
Don't need this in the template context you can embed the text here.

dmitryax replied:
I moved to a const just to use in tests for generated file validation. Just don't like to repeat it there. May we be can keep it?

dmitryax replied:
I just want to make sure that the correct file is generated based on the flag, so I'm using this line as anchor. Let me know if you want me to duplicate this string in tests. Or I can generate a file with different name like `genrerate_metrics_v2.go`

bogdandrutu replied:
generate file with different is actually better I think, since that way we may even be able to have both in parallel.

dmitryax replied:
Do we need to have both in parallel? I think we should keep only one to make sure the old functionality is not being used be scraper

dmitryax replied:
Moved it to `generate_metrics_v2.go` but removed the old one.

dmitryax replied:
But now at least it's possible to have both enabled. If we ever need that, we can change the generator command to support that

dmitryax commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml:
```diff
@@ -10,10 +10,12 @@ attributes:
       aggregation: cumulative
       monotonic: true
-    attributes: [state]
+    attributes: [cpu, state]
```
This is more like unrelated but required a fix. I'll move it to another PR

dmitryax replied:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6422

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go:
```diff
@@ -44,13 +44,21 @@ func newCPUScraper(_ context.Context, cfg *Config) *scraper {
 
+// Assumed number of CPU cores
+// TODO: use gopsutil to get the real number
+const cpuCoresNumber = 16
```
runtime.NumCPUs?

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go:
```diff
@@ -44,13 +44,21 @@ func newCPUScraper(_ context.Context, cfg *Config) *scraper {
-	s.startTime = pdata.Timestamp(bootTime * 1e9)
+	withStartTime := metadata.WithStartTime(pdata.Timestamp(bootTime * 1e9))
+	withStatesCapacity := metadata.WithAttributeStateCapacity(cpuStatesLen)
+	withCoresCapacity := metadata.WithAttributeStateCapacity(cpuCoresNumber)
```
Should this be WithAttributeCpu...?

dmitryax replied:
Right! Thanks for catching it

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go:
```diff
@@ -64,31 +72,21 @@ func (s *scraper) scrape(_ context.Context) (pdata.Metrics, error) {
+	value float64) {
 	// ignore cpu attribute if reporting "total" cpu usage
+	cpuAttributeValue := pdata.NewAttributeValueEmpty()
 	if cpuLabel != gopsCPUTotal {
```
Do we know when this happens?

dmitryax replied:
Yes, it's always reported on windows -- no per CPU metrics there. Also gopsutil has on option to not peport per CPU metrics, in that case this attribute is reported

bogdandrutu replied:
I don't believe you, see https://github.com/shirou/gopsutil/blob/a00b8eac527dd3e883ed6460ca1edc0915bbb6cb/v3/cpu/cpu_windows.go#L63 

I think this is reported only if "Times" func is called with perCpu = false.

dmitryax replied:
Oh you right, I missed that. 

Do you think we should remove the functionality to skip attributes with the `NewAttributeValueEmpty()` until we have a need for that?

dmitryax replied:
I removed it for now and set "string" type to be expected in the helper interface. We can change it later once other attribute types are supported.

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go:
```diff
@@ -64,31 +72,21 @@ func (s *scraper) scrape(_ context.Context) (pdata.Metrics, error) {
 
-func initializeCPUTimeDataPoint(dataPoint pdata.NumberDataPoint, startTime, now pdata.Timestamp, cpuLabel string, stateLabel string, value float64) {
-	attributes := dataPoint.Attributes()
+func initializeCPUTimeDataPoint(mb *metadata.MetricsBuilder, now pdata.Timestamp, cpuLabel string, stateLabel string,
```
initializeCPUTimeDataPoint/recordSystemCPUTimeDataPoint? or just `recordCPUTimeDataPoint`

dmitryax replied:
changed

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go:
```diff
@@ -26,13 +26,13 @@ import (
-	initializeCPUTimeDataPoint(ddps.AppendEmpty(), startTime, now, cpuTime.CPU, metadata.AttributeState.Softirq, cpuTime.Softirq)
-	initializeCPUTimeDataPoint(ddps.AppendEmpty(), startTime, now, cpuTime.CPU, metadata.AttributeState.Steal, cpuTime.Steal)
-	initializeCPUTimeDataPoint(ddps.AppendEmpty(), startTime, now, cpuTime.CPU, metadata.AttributeState.Wait, cpuTime.Iowait)
+func appendCPUTimeStateDataPoints(mb *metadata.MetricsBuilder, now pdata.Timestamp, cpuTime cpu.TimesStat) {
```
same probably use "record" instead of append.

dmitryax replied:
changed

dmitryax committed:
Apply new experimental metrics builder for hostmetrics/cpu scraper

Introduce new experimental metrics builder interface for metrics scrapers and apply it to hostmetrics/cpu scraper.

djaglowski commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go:
```diff
@@ -50,7 +51,11 @@ func (s *scraper) start(context.Context, component.Host) error {
+	withStartTime := metadata.WithStartTime(pdata.Timestamp(bootTime * 1e9))
+	withStatesCapacity := metadata.WithAttributeStateCapacity(cpuStatesLen)
+	withCoresCapacity := metadata.WithAttributeCpuCapacity(runtime.NumCPU())
+	s.mb = metadata.NewMetricsBuilder(s.config.Metrics, withStartTime, withStatesCapacity, withCoresCapacity)
```
Some scrapers will have many more "Capacity" parameters. Should we render these into a separate functions, perhaps like this?
```go
s.mb = metadata.NewMetricsBuilder(s.config.Metrics, withStartTime).
                WithStatesCapacity(withStatesCapacity).
                WithCoresCapacity(withCoresCapacity)
```

dmitryax replied:
In that case I would change it for all including withStartTime. I like it, but seems like it's not very common pattern in the collector's codebase. @bogdandrutu what's your thought on this?

bogdandrutu replied:
I think this is the most "go" way to have this:

```golang
s.mb = metadata.NewMetricsBuilder(
	s.config.Metrics,
	metadata.WithStartTime(pdata.Timestamp(bootTime * 1e9)),
	metadata.WithAttributeStateCapacity(cpuStatesLen),
	metadata.WithAttributeCpuCapacity(runtime.NumCPU()))
```

djaglowski commented on receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go:
```diff
@@ -64,31 +69,21 @@ func (s *scraper) scrape(_ context.Context) (pdata.Metrics, error) {
 
-func initializeCPUTimeDataPoint(dataPoint pdata.NumberDataPoint, startTime, now pdata.Timestamp, cpuLabel string, stateLabel string, value float64) {
-	attributes := dataPoint.Attributes()
+func recordCPUTimeDataPoint(mb *metadata.MetricsBuilder, now pdata.Timestamp, cpuAttribute string,
+	stateLabel string, value float64) {
```
Maybe just personal preference, but I would encourage that the metrics builder is not passed around, that instead this is a receiver function:

```suggestion
func (s *scraper) recordCPUTimeDataPoint(now pdata.Timestamp, cpuAttribute string, stateLabel string, value float64)
	...
	s.mb.RecordSystemCPUTimeDataPoint(...)
```

What I like about this pattern is that the parameters to the function are purely descriptive of the data point, and the concern of "how to build the data point" is assigned to the scraper.

Ultimately, this is just an individual implementation detail, but we are encouraging a pattern to reuse many times, so I think it's worth considering here.

dmitryax replied:
Agree, makes sense to establish this pattern. I'll change it

bogdandrutu commented on cmd/mdatagen/metrics_v2.tmpl:
```diff
@@ -0,0 +1,211 @@
+func (mb *MetricsBuilder) Emit(metrics pdata.MetricSlice) {
+	{{- range $name, $metric := .Metrics }}
+	if mb.config.{{- $name.Render }}.Enabled {
+		mb.metrics.{{- $name.Render }}.CopyTo(metrics.AppendEmpty())
```
MoveTo?
